### PR TITLE
[T22] Libreplex royalty enforcement support

### PR DESCRIPTION
**Pull request by solonk8**
Support libreplex royalty standard for both legacy and latest royalty structure https://libreplex-1.gitbook.io/libreplex/royalty-enforcement

1. royalty is added to this two entry points:
- sol_ext_fulfill_buy
- sol_ext_fulfill_sell
2. sdk integration to support:
- royalty enforcement
- shared escrow remaining account

tested for buy/sell/both sides as well as shared escrow

integration test screenshot
deposit sell:
deposit-sell

withdraw sell:
withdraw-sell

fulfill buy:
fulfill-buy

fulfill buy with shared escrow:
fulfill-buy-shared-escrow

fulfill sell:
fulfill-sell

### CODE REVIEW

File: Anchor.toml [hunk 1]
```diff
@@ -30,6 +30,12 @@ address = "ocp4vWUzA2z2XMYJ3QhM9vWdyoyoQwAFJhRdVTbvo9E" # ocp: open_creator_prot
 [[test.validator.clone]]
 address = "6Huqrb4xxmmNA4NufYdgpmspoLmjXFd3qEfteCddLgSz" # ocp: policy (allow all)
 
+[[test.validator.clone]]
+address = "99jtJwGDfaBKXtc7kxQneAGbERGK8F5XyJWHv7qTbj9G" # global deny list for libreplex royalty enforcement
+
+[[test.validator.clone]]
```
Comment (solonk8) [inline Anchor.toml:36]: @madergaser I was wrong before that cloning does not work... I guess I messed up with my local somehow

File: programs/mmm/src/instructions/ext_vanilla/sol_ext_fulfill_buy.rs [hunk 2]
```diff
@@ -261,13 +269,30 @@ pub fn handler<'info>(
         ))?;
     }
 
+    let mut royalty_paid: u64 = 0;
```
Comment (nothing0012) [inline programs/mmm/src/instructions/ext_vanilla/sol_ext_fulfill_buy.rs:272]: we can simplify the conditions here

File: programs/mmm/src/util.rs [hunk 3]
```diff
@@ -450,6 +451,63 @@ pub fn get_metadata_royalty_bp(
...
+        anchor_lang::solana_program::program::invoke_signed(
+            &anchor_lang::solana_program::system_instruction::transfer(
+                payer.key,
+                creator_account.key,
+                royalty,
+            ),
+            &[
+                payer.to_account_info(),
+                creator_account.to_account_info(),
+                system_program.to_account_info(),
```
Comment (madergaser) [inline programs/mmm/src/util.rs:503]: nit: no need for system_program

File: sdk/src/transferHookProvider.ts [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+  getMint,
+  getTransferHook,
+  Mint,
+  resolveExtraAccountMeta,
+  TOKEN_2022_PROGRAM_ID,
+} from '@solana/spl-token';
+import { unpack } from '@solana/spl-token-metadata';
+import { AccountMeta, Connection, PublicKey } from '@solana/web3.js';
+
+export interface TransferHookProvider {
```
Comment (nothing0012) [inline sdk/src/transferHookProvider.ts:16]: what we need is a central logic of `TransferHookProviderSelector`, and the goal can be minimizing the rpc requests

File: tests/utils/generic.ts [hunk 5]
```diff
@@ -19,6 +19,7 @@ export const PRICE_ERROR_RANGE = 50;
 export const OCP_COMPUTE_UNITS = 1_400_000;
 export const MIP1_COMPUTE_UNITS = 700_000;
 export const IMMUTABLE_OWNER_EXTENSION_LAMPORTS = 34_800;
+export const TRANSFER_HOOK_COMPUTE_UNITS = 1_400_000;
```
Comment (nothing0012) [inline tests/utils/generic.ts:22]: relax this number to 800k

**Review by nothing0012 — APPROVED** [review]

File: programs/mmm/src/util.rs [hunk 6]
```diff
@@ -450,6 +451,59 @@ pub fn get_metadata_royalty_bp(
...
 }
 
+#[allow(clippy::too_many_arguments)]
+pub fn pay_creator_fees_in_sol_ext<'info>(
+    total_price: u64,
+    optional_creator_account: Option<&AccountInfo<'info>>,
+    payer: AccountInfo<'info>,
+    sfbp: u16,
+    payer_seeds: &[&[&[u8]]],
+    system_program: AccountInfo<'info>,
```
Comment (madergaser) [inline programs/mmm/src/util.rs:461]: nit: unused?

File: sdk/src/transferHookProvider.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        isSigner: false,
+        isWritable: true,
+      },
+      ...previousMetas,
+    ];
+  }
+
+  isTransferHookProgramAllowed(): boolean {
+    return this.transferHookProgramId
+      ? ALLOWED_TRANSFER_HOOK_PROGRAM_IDS.includes(this.transferHookProgramId)
```
Comment (madergaser) [inline sdk/src/transferHookProvider.ts:125]:
```suggestion
      ? ALLOWED_TRANSFER_HOOK_PROGRAM_IDS.some((v) => v.equals(this.transferHookProgramId))
```

File: tests/utils/nfts.ts [hunk 8]
```diff
@@ -225,13 +387,53 @@ export async function createTestMintAndTokenT22VanillaExt(
     tokenProgramId,
   );
 
+  const ixs: TransactionInstruction[] = [];
+  if (
+    transferHookArgs?.transferHookProgramId ===
```
Comment (madergaser) [inline tests/utils/nfts.ts:392]: .equals ?

**Review by madergaser — APPROVED** [review]

**Pull request merged**